File: examples/Node.js/formSubmissionDecrypter.ts
import { privateDecrypt, createDecipheriv } from "node:crypto";
import type {
  EncryptedFormSubmission,
  PrivateApiKey,
} from "./dataStructures.js";

export function decryptFormSubmission(
  encryptedSubmission: EncryptedFormSubmission,
  privateApiKey: PrivateApiKey,
): string {
  const privateKey = {
    key: privateApiKey.key,
    oaepHash: "sha256",
  };

  const decryptedKey = privateDecrypt(
    privateKey,
    Buffer.from(encryptedSubmission.encryptedKey, "base64"),
  );

  const decryptedNonce = privateDecrypt(
    privateKey,
    Buffer.from(encryptedSubmission.encryptedNonce, "base64"),
  );

  const decryptedAuthTag = privateDecrypt(
    privateKey,
    Buffer.from(encryptedSubmission.encryptedAuthTag, "base64"),
  );

  const gcmDecipher = createDecipheriv(
    "aes-256-gcm",
    decryptedKey,
    decryptedNonce,
  );

  gcmDecipher.setAuthTag(decryptedAuthTag);

  const encryptedData = Buffer.from(
    encryptedSubmission.encryptedResponses,
    "base64",
  );

  const decryptedData = Buffer.concat([
    gcmDecipher.update(encryptedData),
    gcmDecipher.final(),
  ]);

  return decryptedData.toString("utf8");
}

File: examples/.NET/FormSubmissionDecrypter.cs
using System.Security.Cryptography;
using System.Text;

namespace dotnet
{
  public class FormSubmissionDecrypter
  {
    public static string Decrypt(EncryptedFormSubmission encryptedSubmission, PrivateApiKey privateApiKey)
    {
      try
      {
        RSA rsa = RSA.Create();

        rsa.ImportFromPem(privateApiKey.key);

        byte[] decryptedKey = rsa.Decrypt(Convert.FromBase64String(encryptedSubmission.encryptedKey), RSAEncryptionPadding.OaepSHA256);
        byte[] decryptedNonce = rsa.Decrypt(Convert.FromBase64String(encryptedSubmission.encryptedNonce), RSAEncryptionPadding.OaepSHA256);
        byte[] decryptedAuthTag = rsa.Decrypt(Convert.FromBase64String(encryptedSubmission.encryptedAuthTag), RSAEncryptionPadding.OaepSHA256);

        byte[] encryptedData = Convert.FromBase64String(encryptedSubmission.encryptedResponses);

        AesGcm aesGcm = new(decryptedKey, 16);

        byte[] decryptedData = new byte[encryptedData.Length];

        aesGcm.Decrypt(decryptedNonce, encryptedData, decryptedAuthTag, decryptedData);

        return Encoding.UTF8.GetString(decryptedData);
      }
      catch (Exception exception)
      {
        throw new Exception("Failed to decrypt form submission", exception);
      }
    }
  }
}